### PR TITLE
Add Service API's into OSD(#425)

### LINKED ISSUE #425: Feature: Service API (Part 1) [WIP]
**Is this a BUG REPORT or FEATURE REQUEST?**:
FEATURE

**Reason Needed**:
To create OSD-SDK
Move internal PX API to OSD to create external API SDK.

# Service API (Part 1)
Service level commands allows maintenance related operation of drives and drive pools. The most common cases would be for Disk addition/replacement. Here are all suppported API's -

 # a. Enter Maintenance mode
Enter Maintenance mode take node out of cluster and allows user to perform physical maintenance of node.
    
* **Method** : PUT
* **Endpoint** : /entermaintenance
* **Content-type** : application/json
* **Response status** : 200

**Example Request** :
```PUT /v1/service/entermaintenance```
**Example Response**:
200

# b. Exit Maintenance mode
Exit maintenance mode will put node back into cluster
* **Method** : PUT
* **Endpoint** : /exitmaintenance
* **Content-type** : application/json
* **Response status** : 200

**Example Request** :
```PUT /v1/service/exitmaintenance```
**Example Response** :
200

# c. Add Drive 
  Add drive to cluster storage pool
* **Method** : POST
* **Endpoint** : /drive
* **Content-type** : application/json
* **Response status** : 200
* **Params** :
   operation (string) : start/status of drive add operation (default start)
   drive (string) : drive to add e.g /dev/sda, /dev/sdb
   journal(bool) : Use this drive as a journal device

**Example Request**
```POST /v1/service/drive```
```sh
Content-type : application/json
{
    drive : "/dev/sda",
    journal : true
}
```
**Example Response** :
200

# d. Replace Drive
Let you replace old drive with new drive.
* **Method** : PUT
* **Endpoint** :  /drive
* **Content-type** : application/json
* **Response status** : 200
* **Params** :
    operation(string): start/status of replace operation
    source (string) : source drive to replace
    target(string) : target drive to replace source drive with

**Example Request** :
```PUT /v1/service/drive```
```sh
Content-type : application/json
{
    source : "/dev/sda",
    target : "/dev/sdb"
}
```
**Example Response** :
200

# e. Rebalance the storage pool
Pool rebalance does spread data across all available drives in the pool.
* **Method** : PUT
* **Endpoint** :  /rebalancepool
* **Content-type** : application/json
* **Response status** : status(string)
* **Params** :
    operation(string): start/status of  operation
   poolID (string): poolID to rebalance

**Example Request** :
```PUT /v1/service/rebalancepool```
```sh
Content-type : application/json
{
   operation : "start",
   poolID : "0"
}
```
**Example Response** :
200
```sh
{
    status : "Running"
}
```

**Interface** : 

```sh
type service interface {
   // ServiceAddDrive adds the specified drive
    ServiceAddDrive(op, drive string, journal bool) (string, error)

    // ReplaceDrive source with target.
    ServiceReplaceDrive(op, source, target string) (string, error)

    // Rebalance the storage pool
    ServiceRebalancePool(op string, poolID int) (string, error)

    // ExitMaintenanceMode exits maintenance mode
    ServiceExitMaintenanceMode() error

    // EnterMaintenanceMode enters maintenance mode and exits the proces
    ServiceEnterMaintenanceMode(exitOut bool) error
}
```

### PR DESCRIPTION
**Pull request by ram-infrac**
Signed-off-by: Ram <ram.suradkar@portworx.com>

**What this PR does / why we need it**:
This is PR for #425 , which will add Service API endpoints in OSD with default ```Not implemented``` status.

**Which issue(s) this PR fixes** (optional)
Closes #425

**Special notes for your reviewer**:

### CODE REVIEW

File: api/server/services.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+	status, err := c.ServiceManager.ServiceRebalancePool(srvReq.Operation, srvReq.PoolID)
+	if err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	srvResp := &services.ServiceMessage{
+		Status:  status,
+		Version: "v1",
```
Comment (ram-infrac) [inline api/server/services.go]: Should it be clusterAPIVersion?

File: services/services.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	ServiceReplaceDrive(op, source, target string) (string, error)
+
+	// ServiceRebalance the storage pool
+	ServiceRebalancePool(op string, poolID int) (string, error)
+
+	// ServiceExitMaintenanceMode exits maintenance mode
+	ServiceExitMaintenanceMode() error
+
+	// ServiceEnterMaintenanceMode enters maintenance mode and exits the proces
+	ServiceEnterMaintenanceMode(exitOut bool) error
```
Comment (ram-infrac) [inline services/services.go]: @prabirpaul  do we need exitOut as argument here?

File: api/client/cluster/services.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
+package cluster
+
+import (
+	"github.com/libopenstorage/openstorage/services"
+)
+
+// ServiceEnterMaintenanceMode takes node out of cluster
+func (c *clusterClient) ServiceEnterMaintenanceMode(exitOut bool) error {
```
Comment (lpabon) [inline api/client/cluster/services.go]: What is exitOut? Probably not needed right?

File: api/client/cluster/services.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	res := req.Do()
+	if res.Error() != nil {
+		return res.FormatError()
+	}
+
+	return nil
+}
+
+// ServiceAddDrive adds specified drive
+func (c *clusterClient) ServiceAddDrive(op, drive string, journal bool) (string, error) {
```
Comment (lpabon) [inline api/client/cluster/services.go]: What is op? Is it needed?

File: api/client/cluster/services.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	req := c.c.Put().Resource(clusterPath + ServicePath + "/drive").Body(request)
+	if err := req.Do().Unmarshal(&srvResp); err != nil {
+		return srvResp.Status, err
+	}
+
+	return srvResp.Status, nil
+}
+
+// ServiceRebalancePool reblance the storage pool
+func (c *clusterClient) ServiceRebalancePool(op string, poolID int) (string, error) {
```
Comment (lpabon) [inline api/client/cluster/services.go]: All other APIs make sense, but this one is a little weird. May want to add in the documentation that the poolID refers to api.proto->StoragePool, right?

File: api/server/cluster.go [hunk 6]
```diff
@@ -13,6 +13,7 @@ import (
 	"github.com/libopenstorage/openstorage/cluster"
 	sched "github.com/libopenstorage/openstorage/schedpolicy"
 	"github.com/libopenstorage/openstorage/secrets"
+	serv "github.com/libopenstorage/openstorage/services"
```
Comment (lpabon) [inline api/server/cluster.go]: Just curious, right just `serv` not `services`?

File: api/server/services.go (hunk 1 above)
Comment (lpabon) [reply api/server/services.go]: Is this needed? Why is the version there?

File: api/client/cluster/services.go (hunk 3 above)
Comment (ram-infrac) [reply api/client/cluster/services.go]: right, I commented about same above^

File: api/client/cluster/services.go (hunk 4 above)
Comment (ram-infrac) [reply api/client/cluster/services.go]: it's for start operation/status of current operation

**Pull request closed without merging**